src/MongoDB/CursorInterface.c

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <php.h>
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+
+zend_class_entry* php_phongo_cursor_interface_ce;
+
+/* {{{ MongoDB\BSON\CursorInterface function entries */
+ZEND_BEGIN_ARG_INFO_EX(ai_CursorInterface_setTypeMap, 0, 0, 1)
+	ZEND_ARG_ARRAY_INFO(0, typemap, 0)
+ZEND_END_ARG_INFO()
+
+ZEND_BEGIN_ARG_INFO_EX(ai_CursorInterface_void, 0, 0, 0)
+ZEND_END_ARG_INFO()
+
+static zend_function_entry php_phongo_cursor_interface_me[] = {
+	/* clang-format off */
+	ZEND_ABSTRACT_ME(CursorInterface, getId, ai_CursorInterface_void)
+	ZEND_ABSTRACT_ME(CursorInterface, getServer, ai_CursorInterface_void)
+	ZEND_ABSTRACT_ME(CursorInterface, isDead, ai_CursorInterface_void)
+	ZEND_ABSTRACT_ME(CursorInterface, setTypeMap, ai_CursorInterface_setTypeMap)
+	ZEND_ABSTRACT_ME(CursorInterface, toArray, ai_CursorInterface_void)
+	PHP_FE_END
+	/* clang-format on */
+};
+/* }}} */
+
+void php_phongo_cursor_interface_init_ce(INIT_FUNC_ARGS) /* {{{ */
+{
+	zend_class_entry ce;
+
+	INIT_NS_CLASS_ENTRY(ce, "MongoDB\\Driver", "CursorInterface", php_phongo_cursor_interface_me);
+	php_phongo_cursor_interface_ce = zend_register_internal_interface(&ce TSRMLS_CC);
```

Comment:
In order for this to fully stand in for Cursor, I think it needs to extend Traversable. This was discussed in my original comment that resulted in PHPC-1123: https://github.com/mongodb/mongo-php-driver/issues/754#issuecomment-365683894

With that change, I think we can then allow Cursor to implement Traversable through CursorInterface, instead of implementing both directly.

Reply:
Done. In fact, if we add ``zend_class_implements`` to CursorInterface, we *cannot* have it in Cursor:

> ``E_COMPILE_ERROR: Class MongoDB\Driver\CursorInterface cannot implement previously implemented interface Traversable``

Reply:
Interesting. Is this something that only applies to PHP internals, or would this affect userland?

https://3v4l.org/OhPT2 suggests that it's just an internal restriction. And I wouldn't expect there to be any any issue with a class implementing both Iterator and IteratorAggregate (in userland).